src/test/service/GooglePlaceService.test.ts

```diff
@@ -0,0 +1,109 @@
+import { Client, Status } from '@googlemaps/google-maps-services-js';
+import axios from 'axios';
+import { CustomAxiosMockAdapter } from '../mock/CustomAxiosMockAdapter';
+import { GooglePlaceService } from '../../main/service/GooglePlaceService';
+import { PlaceDetailsResponseData } from '@googlemaps/google-maps-services-js/dist/places/details';
+
+/**
+ * We are using MockAdapter to mock axios requests made by Client. This is because for tests
+ * we do not want to make actual requests to Google Places API.
+ * NOTE: if you do want to test with actual google API and not the mock
+ * 1. import dotenv and call dotenv.config()
+ * 2. Remove all usages of the mock adapter
+ * NOTE: Here is an example of a placeId which will result in NOT_FOUND
+ * * placeId: EjhDdW1odXJpeWV0IE1haGFsbGVzaSwgVXp1biBTay4sIEV6aW5lL8OHYW5ha2thbGUsIFR1cmtleSIuKiwKFAoSCakQkmN8XrAUEVkLpNK_F4IJEhQKEgmFzKyYe16wFBGSjU7F2ooIIg
+ * * source: https://stackoverflow.com/questions/57261582/google-places-api-refresh-place-ids-with-java
+ * * When used in a refresh request, it will return a new valid placeId.
+ * */
+describe('Unit tests for GooglePlaceService', function () {
+  const SAMPLE_PLACE_ID = 'ChIJb0n5cWl3hlQRIbVGYLiTEgE';
+  let googlePlaceService: GooglePlaceService;
+  let customAxiosMockAdapter: CustomAxiosMockAdapter;
+
+  beforeEach(() => {
+    const axiosInstance = axios.create();
+    googlePlaceService = new GooglePlaceService(new Client({ axiosInstance }));
+    customAxiosMockAdapter = new CustomAxiosMockAdapter(axiosInstance);
+  });
+
+  test('Should be able to successfully get place details', async () => {
+    customAxiosMockAdapter.mockSuccessfulPlaceDetails(SAMPLE_PLACE_ID);
+    try {
+      const placeDetailsResponseData: PlaceDetailsResponseData = await googlePlaceService.getRestaurantDetails(
+        SAMPLE_PLACE_ID
+      );
+      expect(placeDetailsResponseData.status).toEqual('OK');
+      expect(placeDetailsResponseData.result.place_id).toEqual(SAMPLE_PLACE_ID);
+      expect(placeDetailsResponseData.result.name).toEqual('MOCK NAME');
+    } catch (e) {
+      fail('Should not have failed: ' + e);
+    }
+  });
+
+  test('If response results in INVALID_REQUEST, no data should be present', async () => {
+    customAxiosMockAdapter.mockInvalidRequestPlaceDetails();
```

Comment:
Just to help me better understand the mock, how do these calls ensure it's using the mocked version?
ie `customAxiosMockAdapter.mockInvalidRequestPlaceDetails()`?

Reply:
good question. 

First, we need to configure the mocks. We're making sure that `Client` and `CustomAxiosMockAdapter` share the same `axiosInstance`, otherwise this wouldn't work.
https://github.com/ubclaunchpad/PromoPal-backend/blob/6886d32e9dc59757253105162479a46a6d756cfe/src/test/service/GooglePlaceService.test.ts#L23-L27

By saying `customAxiosMockAdapter.mockInvalidRequestPlaceDetails()`, we're telling the mock adapter that on any axios request that hits the specified endpoint, we will return an invalid request response. Since `Client` uses `axios` to make the request, we've mocked axios so instead of making an http request, we'll return what the mock has configured.

Reply:
Oooh, gotcha! Thanks for explaining 😄